contracts/BlockDecoder.sol

```diff
@@ -0,0 +1,380 @@
+    using RLPReader for bytes;
+    using Strings for uint256;
+
+    struct Header {
+        bytes Bloom; //[256]byte
+        HashData hashData;
+        BaseData baseData;
+        ValidatorData validatorData;
+        Commit commit;
+    }
+
+    struct HashData {
+        bytes32 ParentHash;
+        bytes32 UncleHash;
+        address Coinbase;
+        bytes32 Root;
+        bytes32 TxHash;
+        bytes32 ReceiptHash;
+    }
+
+    struct BaseData {
+        uint256 Difficulty;
+        uint256 Number;
+        uint64 GasLimit;
+        uint64 GasUsed;
+        uint64 Time;
+        bytes Extra;
+        bytes32 MixDigest;
+        bytes8 Nonce;
+        uint256 BaseFee;
+    }
+
+    struct ValidatorData {
+        uint64 TimeMs;
+        address[] NextValidators;
+        uint256[] NextValidatorPowers;
+        bytes32 LastCommitHash;
+    }
+
+    struct Commit {
+        uint64 Height;
+        uint32 Round;
+        bytes32 BlockID;
+        CommitSig[] Signatures;
+    }
+
+    struct CommitSig {
+        uint8 BlockIDFlag;
+        address ValidatorAddress;
+        uint64 TimestampMs;
+        bytes Signature; //[R || S || V]
+    }
+
+    struct voteForSign {
+        SignedMsgType Type;
+        uint64 Height;
+        uint32 Round;
+        bytes32 BlockID;
+        uint64 TimestampMs;
+        string ChainID;
+    }
+
+    uint8 constant BLOCK_FLAG_COMMIT = 2;
+
+    enum SignedMsgType {
+        UnknownType,
+        // Votes
+        PrevoteType,
+        PrecommitType,
+        // Proposals
+        ProposalType
+    }
+
+    enum HeaderProperty {
+        ParentHash,
+        UncleHash,
+        Coinbase,
+        Root,
+        TxHash,
+        ReceiptHash,
+        Bloom, //[256]byte
+        Difficulty,
+        Number,
+        GasLimit,
+        GasUsed,
+        Time,
+        Extra,
+        MixDigest,
+        Nonce,
+        BaseFee,
+        TimeMs,
+        NextValidators,
+        NextValidatorPowers,
+        LastCommitHash,
+        Commit
+    }
+
+    function verifyHeader(
+        bytes memory headerRlpBytes,
+        bytes memory commitRlpBytes,
+        address[] memory validators,
+        uint256[] memory votePowers
+    )
+        internal
+        pure
+        returns (
+            bool,
+            uint256,
+            bytes32
+        )
+    {
+        // ToDo:verify header base data
+        bytes32 hash = msgHash(headerRlpBytes);
+        uint256 height = decodeHeaderHeight(headerRlpBytes);
+
+        Commit memory commit = decodeCommit(commitRlpBytes.toRlpItem());
+        require(commit.BlockID == hash, "incorrect BlockID");
+        require(commit.Height == height, "incorrect Height");
+
+        // verify all signatures
+        require(
+            verifyAllSignature(commit, validators, votePowers, true, false, votingPowerNeed(votePowers), 3334),
+            "failed to verify all signatures"
+        );
+
+        return (true, height, hash);
```

Comment:
Looks like (false, ..., ...) will never be returned?

We should either stick to either returning true/false or require(...).